_posts/2020-06-10-connect-kafka-spring-kafka.md

```diff
@@ -0,0 +1,575 @@
+
+### Message Listener
+
+As said earlier, `KafkaMessageListenerContainer` receives all records from all topics on a single thread. The `ConcurrentMessageListenerContainer` assigns these records to multiple `KafkaMessageListenerContainer` instances to provide the multi-threaded capability. It also has a concurrency property (`setConcurrency`) which helps to instantiate multiple `KafkaMessageListenerContainer` instances. Let's now look at how to create listeners:
+
+### Using `@KafkaListener` Annotation
+
+`@KafkaListener` annotation allows us to create listeners effortlessly as shown below:
+
+```java
+@Component
+class KafkaListenersExample {
+
+    Logger LOG = LoggerFactory.getLogger(KafkaListenersExample.class);
+
+    @KafkaListener(topics = "reflectoring-1")
+    void listener(String data) {
+        LOG.info(data);
+    }
+
+    @KafkaListener(topics = "reflectoring-1, reflectoring-2", groupId = "reflectoring-group-2")
+    void commonListenerForMultipleTopics(String message) {
+      LOG.info("MultipleTopicListener - {}", message);
+    }
+}
+```
+
+To use this annotation we should add `@EnableKafka` annotation on one of our `@Configuration` class and it requires a listener container factory, which we have configured in `KafkaConsumerConfig.java`.
+We can create a listener bean by annotating a method with `@KafkaListener` annotation and the created bean will be wrapped in `MessagingMessageListenerAdapter`. We can also specify multiple topics for a single listener using topics attribute as shown above.
+
+### Using `@KafkaHandler` for Class Level
+
+We can use `@KafkaListener` annotation at class-level. If we do so, we should specify `@KafkaHandler` at the method level:
+
+```java
+@Component
+@KafkaListener(id = "class-level", topics = "reflectoring-3")
+class KafkaClassListener {
+  ...
+
+  @KafkaHandler
+  void listen(String message) {
+    LOG.info("KafkaHandler[String] {}", message);
+  }
+
+  @KafkaHandler(isDefault = true)
+  void listenDefault(Object object) {
+    LOG.info("KafkaHandler[Default] {}", object);
+  }
+}
+```
+
+When the listener receives records, converted record types are used to determine which `KafkaHandler` must to used. As shown in the above example, records of type `String` will be received by `listen()` and type `Object` will be received by `listenDefault()`. Whenever there is no match default handler will be called and we can assign at most one `KafkaHandler` as default using attribute `isDefault=true`,
+
+### Consuming Records from the Specific Partition with an Initial Offset
+
+We can configure listeners created to listen to multiple topics, partitions, and initial offset optionally (For example: If we want to receive all the records sent to a topic from the time of its creation on application startup we can set the initial offset to zero).
+
+```java
+@Component
+class KafkaListenersExample {
+    ...
+
+    @KafkaListener(topicPartitions = @TopicPartition(topic = "reflectoring-1", partitionOffsets = {
```

Comment:
I manually re-formatted this code to make it better readable.

Reply:
Sorry! I missed it